mnemosyne/doctor.py

```
@@ -0,0 +1,1509 @@
+    constraint behavior.  The caller owns and must close the connection.
+    """
+
+    uri = f"{Path(db_path).absolute().as_uri()}?mode=ro"
+    conn = sqlite3.connect(uri, uri=True)
+    conn.row_factory = sqlite3.Row
+    conn.execute("PRAGMA query_only=ON")
+    conn.execute("PRAGMA foreign_keys=ON")
+    return conn
+
+
+def safe_preview(value: Any, max_length: int = 120) -> str:
```

Ruff (0.15.21)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `value`

(ANN401)